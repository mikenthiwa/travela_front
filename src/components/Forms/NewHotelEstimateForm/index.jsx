import React, { PureComponent } from 'react';
import { PropTypes } from 'prop-types';
import countries from 'world-countries';
import { startCase, toLower } from 'lodash';
import { FormContext, getDefaultBlanksValidatorFor } from '../FormsAPI';
import HotelEstimateFieldset from './FormFieldsets/HotelEstimateDetails';
import SubmitArea from './FormFieldsets/SubmitArea';
import '../NewTravelStipendForm/NewTravelStipendForm.scss';

class NewHotelEstimateForm extends PureComponent {
  constructor(props) {
    super(props);
    this.defaultState = {
      values: {
        travelRegion: '',
        country: '',
        estimate: ''
      },
      errors: {},
      hasBlankFields: true,
      isValidAmount: false
    };
    this.state = { ...this.defaultState };
    this.defaultValidator = getDefaultBlanksValidatorFor(this);
  }
  componentDidMount() {
    this.updateStateData(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      errors: nextProps.editing
        ? nextProps.hotelEstimates.updatedEstimate.errors
        : nextProps.hotelEstimates.error
    });
  }

  updateStateData =  ({
    editing,
    hotelEstimates: { selectedEstimate }
  }) => {
    if (editing && selectedEstimate) {
      const {
        amount: estimate,
        region: travelRegion,
        country
      } = selectedEstimate;
      const regionState = location.search.includes('country=true')
        ? country
        : travelRegion;

      this.setState({
        values: {
          country: country ? regionState : '',
          travelRegion: travelRegion ? regionState : '',
          estimate
        }
      });
    }
  };

  handleSubmit = event => {
    event.preventDefault();
    let { values } = this.state;
    const {
      handleCreateHotelEstimate,
      history,
      closeModal,
      updateHotelEstimate,
      editing,
      location,
      hotelEstimates: {
        selectedEstimate: { id }
      }
    } = this.props;
    const fieldToRemove =
      values.travelRegion === '' ? 'travelRegion' : 'country';
    delete values[fieldToRemove];

    const params = new URLSearchParams(location.search);
    const travelRegionId = params.get('region');
    values['travelRegionId'] = travelRegionId;

    if (editing) {
      updateHotelEstimate(id, values, history);
    } else {
      handleCreateHotelEstimate(values, history);
    }
    closeModal();
  };

  handleShowEventError = event => {
    event.preventDefault();
    this.setState({ isValidAmount: false });
  };

  handleOnchange = event => {
    const {
      target: { value: amount }
    } = event;
    this.setState(
      prevState => {
        return {
          ...prevState,
          values: {
            ...prevState.values,
            estimate: amount
          },
          errors: {},
          isValidAmount: (amount ? (amount > 0 && amount <= 1000) :false),
        };
      },
      this.validate
    );
  };

  getCountryChoices = () => {
    const listOfCountries = countries.map(country => country.name.common);
    const {
      hotelEstimates: { estimates, countriesWithEstimates }
    } = this.props;
    return listOfCountries.filter(
      country => !countriesWithEstimates.includes(country)
    );
  };

  validate = field => {
    const {
      hotelEstimates: { estimates, countriesWithEstimates },
      editing
    } = this.props;

    const regionsWithEstimates = estimates.map(estimate => estimate.region);
    const isValid = this.defaultValidator(field);
    const {
      values: { country, travelRegion }
    } = this.state;
    const countryValue = startCase(toLower(country));
    const regionValue = startCase(toLower(travelRegion));

    this.setState(prevState => ({
      ...prevState,
      errors: {
        ...prevState.errors,
        country:
          countriesWithEstimates.includes(countryValue) && !editing
            ? `A hotel estimate already exists for ${countryValue}.`
            : country.length > 1 &&
              !editing &&
              !this.getCountryChoices().includes(country)
              ? `${country} is not a valid country name.`
              : null,
        travelRegion:
          regionsWithEstimates.includes(regionValue) && !editing
            ? `A hotel estimate already exists for ${regionValue}.`
            : null
      }
    }));
    if ((country.trim().length > 0  && 
    !this.getCountryChoices().includes(country) && !editing) || 
    (travelRegion.trim() == '' && country.trim() == '')) {
      this.setState({ hasBlankFields: true });
    }
    else if (travelRegion == '' && country == '') {
      this.setState({ hasBlankFields: true });
    } else this.setState({ hasBlankFields: false });

    return isValid;
  };

  renderHotelEstimateFieldset = isEmpty => {
    const { values, isValidAmount } = this.state;
    const { editing, hotelEstimates } = this.props;
    const regionfield = location.search.includes('country=true')
      ? ['country', 'filter-dropdown-select']
      : ['travelRegion', 'text'];
    return (
      <HotelEstimateFieldset
        regionfield={regionfield}
        values={values}
        isEmpty={isEmpty}
        amount={values.estimate}
        handleShowEventError={this.handleShowEventError}
        onChangeAmountInput={this.handleOnchange}
        isValidAmount={isValidAmount}
        value="245px"
        editing={editing}
        estimates={hotelEstimates.estimates}
        getCountryChoices={this.getCountryChoices}
      />
    );
  };

  renderForm = () => {
    const { values, hasBlankFields, selection, errors, isValidAmount } = this.state;
    const {
      closeModal,
      hotelEstimates,
      editing,
      updatedEstimate: { isSaving }
    } = this.props;

    let isEmpty;
    if (errors) {
      isEmpty = errors.estimate === 'This field is required';
    }

    return (
      <FormContext
        targetForm={this}
        values={values}
        errors={errors}
        validatorName="validate"
      >
        <form onSubmit={this.handleSubmit} className="new-request">
          {this.renderHotelEstimateFieldset(isEmpty)}
          <SubmitArea
            onCancel={closeModal}
            hotelEstimates={hotelEstimates}
            hasBlankFields={hasBlankFields || !isValidAmount}
            selection={selection}
            loading={isSaving}
            send={editing ? 'Save Estimate' : 'Add Estimate'}
          />
        </form>
      </FormContext>
    );
  };

  render() {
    return <div>{this.renderForm()}</div>;
  }
}

NewHotelEstimateForm.propTypes = {
  closeModal: PropTypes.func,
  hotelEstimates: PropTypes.object,
  handleCreateHotelEstimate: PropTypes.func.isRequired,
  history: PropTypes.object,
  editing: PropTypes.bool,
  updateHotelEstimate: PropTypes.func,
  updatedEstimate: PropTypes.object,
  location: PropTypes.object
};

NewHotelEstimateForm.defaultProps = {
  closeModal: () => {},
  location: {},
  hotelEstimates: {},
  history: {
    push: () => {}
  },
  editing: false,
  updateHotelEstimate: () => {},
  updatedEstimate: {
    isSaving: false
  }
};

export default NewHotelEstimateForm;
