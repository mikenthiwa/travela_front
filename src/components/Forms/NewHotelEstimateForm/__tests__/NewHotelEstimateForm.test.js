import React from 'react';
import NewHotelEstimateForm from '..';

describe('<NewHotelEstimateForm />', () => {
  let wrapper, onSubmit;
  onSubmit = jest.fn();

  const props = {
    location: {
      search: '?country=true',
      pathname: '/travel-cost/hotel-estimates'
    },
    loading: false,
    hotelEstimates: {
      selectedEstimate: { id: 4 },
      isLoading: false,
      estimates: [
        {
          amount: 100,
          country: 'United Kingdom'
        }
      ],
      countriesWithEstimates: ['Kenya', 'Nigeria'],
    },
    hasBlankFields: false,
    user: {
      UserInfo: {
        id: '-LJNw1BsT0LP_E4l2peP',
        name: 'Andrew'
      }
    },
    userData: {
      id: '2',
      fullName: 'Andrew Hinga',
      name: 'Andrew Hinga',
      email: 'andrew.hinga@andela.com',
      userId: '-LJNw1BsT0LP_E4l2peP',
      passportName: 'Andrew Hinga',
      department: 'Fellows',
      occupation: 'Software Developer',
      manager: 'Andrew',
      gender: 'Male',
      createdAt: '2018-09-14T12:48:11.266Z',
      updatedAt: '2018-09-16T07:53:48.835Z',
      roleId: 401938
    },
    userDataUpdate: [],
    handleCreateHotelEstimate: jest.fn(() => {}),
    closeModal: jest.fn(),
    choices: ['Uganda', 'Kenya'],
    modalType: 'create hotel estimate',
    getCountryChoices: jest.fn()
  };

  beforeEach(() => {
    wrapper = mount(<NewHotelEstimateForm {...props} />);
    wrapper.setState({
      hasBlankFields: true,
      errors: { country: '' }
    });
    jest.resetAllMocks();
  });

  it('should call function hasBlankFields', () => {
    const inputField = wrapper.find('.estimate-location').at(0);
    inputField.simulate('change', {
      target: { name: 'country', value: 'ken' }
    });
    expect(wrapper.instance().getCountryChoices());
    const { country: countrErrors } = wrapper.state('errors');
    expect(countrErrors).toEqual('');
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders two fieldsets', () => {
    const fieldsets = wrapper.find('fieldset');
    expect(fieldsets).toHaveLength(2);
  });

  it('renders the hotel estimate fieldset', () => {
    const hotelEstimateDetails = wrapper.find('fieldset .submit-area');
    expect(hotelEstimateDetails).toHaveLength(1);
  });

  it('renders two buttons', () => {
    const buttons = wrapper.find('button');
    expect(buttons).toHaveLength(2);
  });

  it('checks that onSubmit is not called when form is empty', () => {
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  it('renders the hotel estimate amount input without validation error', () => {
    const estimateInput = wrapper.find('input[name="estimate"]');
    const handleOnchange = jest.spyOn(wrapper.instance(), 'handleOnchange');

    wrapper.instance().forceUpdate();

    estimateInput.simulate('change', {
      target: { name: 'estimate', value: '1000' }
    });
    expect(handleOnchange).toHaveBeenCalled();
    expect(wrapper.find('.input-group span.hide-error')).toHaveLength(1);
    expect(wrapper.instance().state.isValidAmount).toBe(true);
  });

  it('renders the hotel estimate amount input with validation error', () => {
    const estimateInput = wrapper.find('input[name="estimate"]');
    const handleOnchange = jest.spyOn(wrapper.instance(), 'handleOnchange');
    const handleShowEventError = jest.spyOn(
      wrapper.instance(),
      'handleShowEventError'
    );

    wrapper.instance().forceUpdate();

    estimateInput.simulate('change', {
      target: { name: 'estimate', value: '-1000' }
    });
    expect(handleOnchange).toHaveBeenCalled();
    estimateInput.simulate('invalid', {
      target: { name: 'estimate', value: '-1000' }
    });
    expect(handleShowEventError).toHaveBeenCalled();
    expect(wrapper.find('.input-group span.show-error')).toHaveLength(1);
    expect(wrapper.instance().state.isValidAmount).toBe(false);
  });

  it('renders the hotel estimate amount input with validation error on `blur`', () => {
    const estimateInput = wrapper.find('input[name="estimate"]');

    wrapper.instance().forceUpdate();

    estimateInput.simulate('blur', { target: { name: 'estimate', value: '' } });
    expect(wrapper.find('.estimate-amount span.error').text()).toBe(
      'This field is required'
    );
  });

  it('should prefill form input fields on editing', () => {
    beforeEach(() => {
      const newProps = {
        ...props,
        editing: true,
        selectedEstimate: {
          id: 5,
          amount: 5,
          region: 'West Africa',
          regionId: 1002
        }
      };
      const wrapper = mount(<NewHotelEstimateForm {...newProps} />);
      jest.resetAllMocks();
    });
    const estimate = wrapper.find('input[name="estimate"]');
    const changeValue = { target: { name: 'estimate', value: '50' } };
    estimate.simulate('change', changeValue);
    const { estimate: estimateValue } = wrapper.state('values');
    expect(estimateValue).toEqual('50');
  });
});
