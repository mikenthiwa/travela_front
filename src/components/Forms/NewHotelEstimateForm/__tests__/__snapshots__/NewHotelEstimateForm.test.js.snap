// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewHotelEstimateForm /> renders correctly 1`] = `
<div>
  <form
    className="new-request"
    key=".0"
    onSubmit={[Function]}
  >
    <fieldset
      className="personal-details"
    >
      <div>
        <div>
          <div
            className="input-group mdl-grid"
          >
            <div
              className="spaces mdl-cell mdl-cell--12-col mdl-cell--12-col-tablet mdl-cell--12-col-phone"
            >
              <div
                className="form-input request{ uniq }dropdown estimate-location "
              >
                <label
                  htmlFor="travelRegion"
                >
                  Type the region
                  <span
                    style={
                      Object {
                        "color": "red",
                      }
                    }
                  >
                     * 
                  </span>
                </label>
                <input
                  autoComplete="off"
                  choices={
                    Array [
                      "Aruba",
                      "Afghanistan",
                      "Angola",
                      "Anguilla",
                      "Åland Islands",
                      "Albania",
                      "Andorra",
                      "United Arab Emirates",
                      "Argentina",
                      "Armenia",
                      "American Samoa",
                      "Antarctica",
                      "French Southern and Antarctic Lands",
                      "Antigua and Barbuda",
                      "Australia",
                      "Austria",
                      "Azerbaijan",
                      "Burundi",
                      "Belgium",
                      "Benin",
                      "Burkina Faso",
                      "Bangladesh",
                      "Bulgaria",
                      "Bahrain",
                      "Bahamas",
                      "Bosnia and Herzegovina",
                      "Saint Barthélemy",
                      "Saint Helena, Ascension and Tristan da Cunha",
                      "Belarus",
                      "Belize",
                      "Bermuda",
                      "Bolivia",
                      "Caribbean Netherlands",
                      "Brazil",
                      "Barbados",
                      "Brunei",
                      "Bhutan",
                      "Bouvet Island",
                      "Botswana",
                      "Central African Republic",
                      "Canada",
                      "Cocos (Keeling) Islands",
                      "Switzerland",
                      "Chile",
                      "China",
                      "Ivory Coast",
                      "Cameroon",
                      "DR Congo",
                      "Republic of the Congo",
                      "Cook Islands",
                      "Colombia",
                      "Comoros",
                      "Cape Verde",
                      "Costa Rica",
                      "Cuba",
                      "Curaçao",
                      "Christmas Island",
                      "Cayman Islands",
                      "Cyprus",
                      "Czechia",
                      "Germany",
                      "Djibouti",
                      "Dominica",
                      "Denmark",
                      "Dominican Republic",
                      "Algeria",
                      "Ecuador",
                      "Egypt",
                      "Eritrea",
                      "Western Sahara",
                      "Spain",
                      "Estonia",
                      "Ethiopia",
                      "Finland",
                      "Fiji",
                      "Falkland Islands",
                      "France",
                      "Faroe Islands",
                      "Micronesia",
                      "Gabon",
                      "United Kingdom",
                      "Georgia",
                      "Guernsey",
                      "Ghana",
                      "Gibraltar",
                      "Guinea",
                      "Guadeloupe",
                      "Gambia",
                      "Guinea-Bissau",
                      "Equatorial Guinea",
                      "Greece",
                      "Grenada",
                      "Greenland",
                      "Guatemala",
                      "French Guiana",
                      "Guam",
                      "Guyana",
                      "Hong Kong",
                      "Heard Island and McDonald Islands",
                      "Honduras",
                      "Croatia",
                      "Haiti",
                      "Hungary",
                      "Indonesia",
                      "Isle of Man",
                      "India",
                      "British Indian Ocean Territory",
                      "Ireland",
                      "Iran",
                      "Iraq",
                      "Iceland",
                      "Israel",
                      "Italy",
                      "Jamaica",
                      "Jersey",
                      "Jordan",
                      "Japan",
                      "Kazakhstan",
                      "Kyrgyzstan",
                      "Cambodia",
                      "Kiribati",
                      "Saint Kitts and Nevis",
                      "South Korea",
                      "Kosovo",
                      "Kuwait",
                      "Laos",
                      "Lebanon",
                      "Liberia",
                      "Libya",
                      "Saint Lucia",
                      "Liechtenstein",
                      "Sri Lanka",
                      "Lesotho",
                      "Lithuania",
                      "Luxembourg",
                      "Latvia",
                      "Macau",
                      "Saint Martin",
                      "Morocco",
                      "Monaco",
                      "Moldova",
                      "Madagascar",
                      "Maldives",
                      "Mexico",
                      "Marshall Islands",
                      "Macedonia",
                      "Mali",
                      "Malta",
                      "Myanmar",
                      "Montenegro",
                      "Mongolia",
                      "Northern Mariana Islands",
                      "Mozambique",
                      "Mauritania",
                      "Montserrat",
                      "Martinique",
                      "Mauritius",
                      "Malawi",
                      "Malaysia",
                      "Mayotte",
                      "Namibia",
                      "New Caledonia",
                      "Niger",
                      "Norfolk Island",
                      "Nicaragua",
                      "Niue",
                      "Netherlands",
                      "Norway",
                      "Nepal",
                      "Nauru",
                      "New Zealand",
                      "Oman",
                      "Pakistan",
                      "Panama",
                      "Pitcairn Islands",
                      "Peru",
                      "Philippines",
                      "Palau",
                      "Papua New Guinea",
                      "Poland",
                      "Puerto Rico",
                      "North Korea",
                      "Portugal",
                      "Paraguay",
                      "Palestine",
                      "French Polynesia",
                      "Qatar",
                      "Réunion",
                      "Romania",
                      "Russia",
                      "Rwanda",
                      "Saudi Arabia",
                      "Sudan",
                      "Senegal",
                      "Singapore",
                      "South Georgia",
                      "Svalbard and Jan Mayen",
                      "Solomon Islands",
                      "Sierra Leone",
                      "El Salvador",
                      "San Marino",
                      "Somalia",
                      "Saint Pierre and Miquelon",
                      "Serbia",
                      "South Sudan",
                      "São Tomé and Príncipe",
                      "Suriname",
                      "Slovakia",
                      "Slovenia",
                      "Sweden",
                      "Eswatini",
                      "Sint Maarten",
                      "Seychelles",
                      "Syria",
                      "Turks and Caicos Islands",
                      "Chad",
                      "Togo",
                      "Thailand",
                      "Tajikistan",
                      "Tokelau",
                      "Turkmenistan",
                      "Timor-Leste",
                      "Tonga",
                      "Trinidad and Tobago",
                      "Tunisia",
                      "Turkey",
                      "Tuvalu",
                      "Taiwan",
                      "Tanzania",
                      "Uganda",
                      "Ukraine",
                      "United States Minor Outlying Islands",
                      "Uruguay",
                      "United States",
                      "Uzbekistan",
                      "Vatican City",
                      "Saint Vincent and the Grenadines",
                      "Venezuela",
                      "British Virgin Islands",
                      "United States Virgin Islands",
                      "Vietnam",
                      "Vanuatu",
                      "Wallis and Futuna",
                      "Samoa",
                      "Yemen",
                      "South Africa",
                      "Zambia",
                      "Zimbabwe",
                    ]
                  }
                  className=" request{ uniq }dropdown estimate-location"
                  label="Type the region"
                  name="travelRegion"
                  onBlur={[Function]}
                  onChange={[Function]}
                  readOnly={false}
                  size=""
                  type="text"
                  value=""
                />
                <span
                  className="error"
                />
              </div>
            </div>
            <div
              className="spaces mdl-cell mdl-cell--12-col mdl-cell--12-col-tablet mdl-cell--12-col-phone"
            >
              <div
                className="form-input request{ uniq }dropdown estimate-amount input-border-error "
              >
                <label
                  htmlFor="estimate"
                >
                  Enter Amount in Dollars ($)
                  <span
                    style={
                      Object {
                        "color": "red",
                      }
                    }
                  >
                     * 
                  </span>
                </label>
                <div>
                  <input
                    autoComplete="off"
                    className=" request{ uniq }dropdown estimate-amount input-border-error"
                    id="your-manager"
                    label="Enter Amount in Dollars ($)"
                    max="1000"
                    min="1"
                    name="estimate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onInvalid={[Function]}
                    placeholder="1000"
                    size=""
                    type="number"
                    value=""
                  />
                  <div
                    className="style-number"
                  >
                    <img
                      alt="icn"
                      className="image1"
                      src="form_select_dropdown.svg"
                    />
                    <img
                      alt="icn"
                      className="image2"
                      src="form_select_dropdown.svg"
                    />
                  </div>
                </div>
                <span
                  className="error"
                />
              </div>
              <span
                className="show-error"
              >
                Amount should be a positive integer and not more than 1000
              </span>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
    <fieldset>
      <div
        className="submit-area"
      >
        <button
          className="bg-btn bg-btn--inactive travel-stipends__right"
          id="cancel"
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="bg-btn bg-btn--active"
          disabled={true}
          id="submit"
          type="submit"
        >
          <undefined />
        </button>
      </div>
    </fieldset>
  </form>
</div>
`;
